File: DSA/RAJAS-BAGGA/Day_9_Question_1.java
class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
    };

    public Node() {
    };
}

class Stack {
    public Node head = null;

    public void push(int data) {
        Node node = new Node(data);
        if (head == null) {
            head = node;
            head.next = null;
            return;
        }
        node.next = head;
        head = node;
    }

    public int pop() {
        if (head == null)
            return -1;
        Node temp = head;
        this.head = head.next;
        int data = temp.data;
        temp = null;
        return data;
    }

    public int getMax() {

        Node tempHead = head;
        int sum = 0;
        while (tempHead != null) {
            sum += tempHead.data;
            tempHead = tempHead.next;
        }
        return sum;
    }

    public void display() {
        if (head == null) {
            System.out.println("Stack empty");
            return;
        }
        Node tempHead = head;
        while (tempHead != null) {
            System.out.println(tempHead.data);
            tempHead = tempHead.next;
        }
    }

}

public class Day_9_Question_1 {
    public static void main(String[] args) {
        Stack stack = new Stack();
        stack.push(0);
        stack.push(1);
        stack.push(2);
        stack.push(3);
        stack.push(4);
        System.out.println(stack.getMax());
    }
}

File: DSA/Pragya_Agarwal/Day9_Question1.py
class MaxStack:
    def __init__(self):
        self.stack = []
        self.max_value = None
    def push(self, x: int)->None:
        if not self.stack:
            self.max_value= x
        else:
            self.max_value =max(x, self.max_value)
        self.stack.append((x, self.max_value))
    def pop(self)->None:
        if self.stack:
            self.stack.pop()
            if self.stack:
                self.max_value =self.stack[-1][1]
            else:
                self.max_value= None
    def get_max(self)->int:
        if self.stack:
            return self.stack[-1][1]
        return None

stack = MaxStack()
stack.push(1)
stack.push(2)
print(stack.get_max())  
stack.pop()
print(stack.get_max())  
